### PR TITLE
Add Anderson acceleration

### PR DESCRIPTION
**Pull request by devmotion**
Fixes https://github.com/JuliaDiffEq/DelayDiffEq.jl/issues/8.

Moreover, this PR updates the type structure and dispatches of the fixed-point iteration (similar to https://github.com/JuliaDiffEq/DiffEqBase.jl/pull/315) and uses `NLFunctional` and `NLAnderson` instead of `FPFunctional` and `FPAnderson`.

### CODE REVIEW

File: src/fpsolve/functional.jl [hunk 1]
```diff
@@ -1,30 +1,52 @@
-function fpsolve!(integrator::DDEIntegrator, fpsolver::FPSolver{<:FPFunctionalConstantCache})
+function fpsolve!(fpsolver::FPSolver{<:Union{NLFunctional,NLAnderson},false},
```
Comment (devmotion) [inline src/fpsolve/functional.jl]: Should we overload `nlsolve!` instead of defining `fpsolve!`?

File: src/fpsolve/type.jl [hunk 2]
```diff
@@ -1,40 +1,34 @@
...
-
-# algorithms
-struct FPFunctional{K,C} <: AbstractFPSolverAlgorithm
-  κ::K
-  fast_convergence_cutoff::C
-  max_iter::Int
+# caches
+struct FPFunctionalCache{uType,uNoUnitsType} <: DiffEqBase.AbstractNLSolverCache
+  du::uType
+  atmp::uNoUnitsType
```
Comment (devmotion) [inline src/fpsolve/type.jl]: Should we also move `du` to the cache in https://github.com/JuliaDiffEq/DiffEqBase.jl/pull/315? And add a unitless cache?

Comment (ChrisRackauckas) [reply src/fpsolve/type.jl]: I think so. It's a bit weird to have something that is alg-specific but then is in the non-specific cache.

File: src/algorithms.jl [hunk 3]
```diff
@@ -17,5 +17,5 @@ If the algorithm is `constrained` only steps of size at most the minimal delay w
 taken. If it is unconstrained, fixed-point iteration `fpsolve` is applied for step sizes
 that exceed the minimal delay.
 """
-MethodOfSteps(alg; constrained = false, fpsolve = FPFunctional()) =
+MethodOfSteps(alg; constrained = false, fpsolve = NLFunctional()) =
```
Comment (ChrisRackauckas) [inline src/algorithms.jl]: I would argue that after this settles we probably want to default to AA. But that doesn't need to happen now.

File: src/fpsolve/functional.jl [hunk 4]
```diff
@@ -40,31 +62,79 @@ function fpsolve!(integrator::DDEIntegrator, fpsolver::FPSolver{<:FPFunctionalCo
...
     if η * ndu < κ && (iter > 1 || iszero(ndu))
       # fixed-point iteration converges
-      fpsolver.status = η < fpsolver.fast_convergence_cutoff ? FastConvergence : Convergence
+      fpsolver.status = η < fast_convergence_cutoff ? FastConvergence : Convergence
       fail_convergence = false
       break
     end
+
+    # perform Anderson acceleration
+    if fpcache isa FPAndersonConstantCache && iter < max_iter
```
Comment (ChrisRackauckas) [inline src/fpsolve/functional.jl]: I did not check this block thoroughly but it looks quite similar to the nlsolve! one so I'm assuming this chunk is good.

File: src/fpsolve/functional.jl (hunk 1 above)
Comment (ChrisRackauckas) [reply src/fpsolve/functional.jl]: ehh, if these are a different cache type then I think we might as well commit to `fpsolve!`

**Pull request merged**